### ISSUE TITLE
Add support for RFC 7239 : Forwarded header

### ISSUE THREAD

**Issue by blop** [haproxy/haproxy#575]
## What should haproxy do differently? Which functionality do you think we should add?

I suggest to add support for RFC 7239 : Forwarded header : https://tools.ietf.org/html/rfc7239

According the the RFC the Forwarded header supports the following case-insensitive parameters separated by a semicolon (and optional space): 
- by : IP address of incoming interface of a proxy
- for : IP address of client making a request through a proxy
- host : Host header field of the incoming request
- proto : Application protocol used for incoming request

People are mostly using other headers like, which are not standard hence the duplicates : 
- X-Forwarded-Proto (http or https)
- X-Forwarded-Host (orignal host)
- X-Forwarded-Server (orignal host)
- X-Forwarded-Port (original host port)
- X-Forwarded-Path (original path before rewrite)
- X-Forwarded-Prefix (rewrite prefix)
- X-Proxied-Https (on/off, true/false)
- X-Real-IP
- X-Client-IP

Currently HAProxy supports the "option forwardfor" which adds a X-Forwarded-For.
But as there is a standardized way, HAProxy should support and encourage it.

**Comment by TimWolla** [comment]
I was already thinking about this, but I also consider `option forwardfor` a configuration flag that is not needed, because to my understanding it's redundant with the `http-request add-header` rules. 

Likewise the `Forwarded` header can already be generated like this (I left out the `by` value, because the RFC recommends using an obfuscated value):

```
	http-request set-var(req.proto) str(https) if { ssl_fc }
	http-request set-var(req.proto) str(http) if ! { ssl_fc }
	http-request add-header Forwarded "for=\"%[src]\"; host=\"%[req.hdr(host)]\"; proto=%[var(req.proto)]"
```

The generation of the `proto` value is a bit ugly. For that a new `choice` converter could be added, returning the first parameter for a true value and the second for a false value: `ssl_fc,choice(https,http)`.

**Comment by quasicomputational** [comment]
There's some annoying subtlety here because the RFC's ABNF doesn't allow for square brackets around IPv4 addresses but they must be present for IPv6, so I think there's value in having one correct implementation in HAProxy rather than N potentially buggy implementations in user configuration.

**Comment by wtarreau** [comment]
I initially wanted to implement "forwarded" until I figured that everyone would want different fields and that it would be easier addressed using add-header. However your point about brackets is perfectly valid, and I think we'd benefit from a converter that takes an address on input and pretty-prints it depending on its family. This would typically add brackets for IPv6.

**Comment by TimWolla** [comment]
> The generation of the `proto` value is a bit ugly. For that a new `choice` converter could be added, returning the first parameter for a true value and the second for a false value: `ssl_fc,choice(https,http)`.

This converter now exists using the name `iif`: 3943e4fc3e8bbfeb5a210c8e86ff0ed53a3bbcc4, leaving just the square bracket issue for IPv6.

**Comment by TimWolla** [comment]
It turns out that adding the square brackets is possible with a pure log-format string. But it definitely is not pretty:

```
http-request set-header forwarded "for=\"%[src,ipmask(32,128),regsub(.*:.*,),length,iif(,[)]%[src,ipmask(32,128)]%[src,ipmask(32,128),regsub(.*:.*,),length,iif(,])]\"; host=\"%[req.hdr(host)]\"; proto=%[ssl_fc,iif(https,http)]"
```

**Comment by wtarreau** [comment]
I think that we need to have an `http_addr()` converter which takes an address on input and emits either a plain IPv4 or a bracketted IPv6. This will significantly simplify everything wherever an address is needed. We can then write `for=%[src,http_addr]` and have exactly the expected output. Tim, are you interested in trying to do it ? It would be a nice addition to your `iif()` in my opinion.

**Comment by TimWolla** [comment]
> Tim, are you interested in trying to do it ?

I don't have a current need for the 'Forwarded' header, but sure, it shouldn't be too hard. However I would like to have this discussed first: https://www.mail-archive.com/haproxy@formilux.org/msg38456.html

On an unrelated note I would like to have a `id=foo` field within the `Forwarded` header to have a standardized place of where to put my unique request IDs, unfortunately there isn't one and proposing that looks like a bureaucratic process :frowning_face:

**Comment by Rawk** [comment]
> On an unrelated note I would like to have a `id=foo` field within the `Forwarded` header to have a standardized place of where to put my unique request IDs, unfortunately there isn't one and proposing that looks like a bureaucratic process frowning_face

@TimWolla: Just a thought. Maybe the node-port can be used (abused?) for unique request IDs. The RFC hints about it:
```
       node-port     = port / obfport
       port          = 1*5DIGIT
       obfport       = "_" 1*(ALPHA / DIGIT / "." / "_" / "-")

   Each of the identifiers may optionally have the port identifier, for
   example, allowing the identification of the endpoint in a NATed
   environment.  The "node-port" can be identified either by its port
   number or by a generated token obfuscating the real port number.  An
   obfuscated port may be used in situations where the possessor of the
   proxy wants the ability to trace requests -- for example, in debug
   purposes -- but does not want to reveal internal information.
```

**Comment by TimWolla** [comment]
@Rawk Yeah, it probably can. However the nice thing about something explicitly standardized would be that it can interoperate with other stuff automatically without requiring manual configuration. For something in-house I can simply continue using `x-req-id` or whatever.

**Comment by waja** [comment]
Hi there,

I wonna just ask, if there was any progress since last year into this?

Many thanks, Jan.

**Comment by Darlelet** [comment]
Hi,

> I think that we need to have an http_addr() converter which takes an address on input and emits either a plain IPv4 or a bracketted IPv6. This will significantly simplify everything wherever an address is needed. We can then write for=%[src,http_addr] and have exactly the expected output. Tim, are you interested in trying to do it ? It would be a nice addition to your iif() in my opinion.

Would a versatile converter directly outputting RFC 7239 compliant node identifiers be a good idea? (1)
```
 IPV4, http_node            -> IPV4
 IPV4, http_node(node-port) -> "IPV4:node-port"
 IPV6, http_node            -> "[IPV6]"
 IPV6, http_node(node-port) -> "[IPV6]:node-port"

 /* extra case: handle "unknown" and obfuscated nodes */
 STR,  http_node            -> "STR"
 STR,  http_node(node-port) -> "STR:node-port"

/* Here node-port is evaluated as a string to allow both numerical ports and obfuscated ports.
 * Using an empty string as node-port argument would result in the
 * same behavior as using http_node converter without argument.
 * /
```

Or should we stick to "basic" converter as suggested by Willy, focusing on IP addresses only, and let the user do the quoting, obfnode and optional port handling if needed? (2)

```
 IPV4, http_addr       -> IPV4
 IPV6, http_addr       -> [IPV6]
```

-----------------------------------

#### With (1), this is how the forwarded header could be generated:

Baseline http-request add-header directive:

```
http-request add-header Forwarded "for=%[var(req.src_addr),http_node(req.src_port)];"
```

for: addr
   ```
   http-request set-var(req.src_addr) src
   ```

   > IPV4 -> forwarded: for=ipv4;
   > IPV6 -> forwarded: for="[ipv6]";

for: addr:port
   ```
   http-request set-var(req.src_addr) src
   http-request set-var(req.src_port) src_port
   ```

   > IPV4 -> forwarded: for="ipv4:port";
   > IPV6 -> forwarded: for="[ipv6]:port";

for: custom_str
   ```
   http-request set-var(req.src_addr) str("custom")
   ```
   > forwarded: for="custom";

 ####  Now with (2):

for: addr
   ```
   http-request add-header Forwarded "for=\"%[src,http_addr]\";"
   ```
   > IPV4 -> forwarded: for="ipv4";
   > IPV6 -> forwarded: for="[ipv6]";

for: addr:port
   ```
   http-request add-header Forwarded "for=\"%[src,http_addr]:%[src_port]\";"
   ```

> IPV4 -> forwarded: for="ipv4:port";
> IPV6 -> forwarded: for="[ipv6]:port";

Known limitation with (2) is that raw ipv4 addresses (without ":node-port") are printed as quoted strings (instead of tokens according to the RFC). But the RFC does not mention whether this is tolerated or not.

--------------------------

To sum up: 
- with (1) and the use of variables: less config duplication (ability to reuse the same add-header directive to handle multiple cases)
- with (2): simpler design but gets complex when multiple cases need to be handled at the same time (requires choosing between multiple add-header directives: error prone) + raw ipv4 address output seems to conflict with RFC

**Comment by wtarreau** [comment]
Thanks Aurélien for your summary. For having given it some thoughts over time, I tend to think that if we want it to be adopted, it should probably be as easy to implement as "option forwardfor", maybe even just an extra argument to that option like we have for a few of them. E.g. "option forwardfor rfc7239". It would make it simple to keep in a defaults section and be used everywhere. However that would mean less ability to select the fields we want to put there. Thus if do that, we'd still need to think a bit how to make it easily extensible so that we can later select what fields to be added there, maybe even using sample expressions, I don't know.

At the very least we need to address what users do with "option forwardfor" nowadays, and natively integrate basic SSL support so that those who currently struggle with x-forwarded-for and x-forwarded-proto can at least be served correctly. There are also a few who use the x-forwarded-host header that contains the original host, though usually it's pointless through a reverse-proxy like haproxy since we're not supposed to rewrite the host header. But some like to have at least the target port the client used (which should normally be in the Host header).

One benefit of relying on an option is that, just like today, we can track the fact that it's been used once and doesn't need to be applied again when switching from a frontend to a backend. Rules make that a bit more complicated in case certain combinations overlap, as they can result in the header being appended twice.

Another point to think about would be the option to sanitize the header and possibly to append to it. Over time, some users have been complaining that we were adding one extra xff instead of appending the same. The issue was obviously on the server behind which didn't properly consider multiple headers, but I understand that appending to an existing one helps with integration. However a year or two ago, some vulnerabilities were reported on XFF processing when quotes were injected. And thanks to our distinct header, we were one of the rare non-vulnerable implementations. So I tend to think that to be nice with users it would be great if we offered the option to append to an existing header, but we should only do it when the current one is valid otherwise we should add a new one. I.e. that can be a later improvement. But I preferred to put all these elements together right now so that they don't get lost.

Also there's something important to take into account (that currently is delaying rfc7239 adoption). It's important that we never pass a request that contains both xff and forwarded, or it creates confusion by carrying two independent lists and again allows clients to inject whatever path they want to. Some choices need to be made here (e.g. prepend xff's contents before forwarded etc). Similarly I think that if there's XFF added by a frontend and forwarded added in the backend, the backend one wins (since it's closer to the server thus more specific). And conversely. That might mean adding that header very late in the processing and always only after the backend switch.

In summary I think we need to double-check what are the common uses right now in 2022, make sure that what we propose is easy and non-ambigous to configure, without introducing security issues. I know it's a lot of (possibly conflicting) requirements but it clearly shows that the design study is way more important than the code at this point.

**Comment by Darlelet** [comment]
You're right the issue is now more than 2 years old, we should re-assess the situation before jumping head first into implementation details, I was a bit rushing here ;).

Thank your for sharing your latest thoughts on this, it is really helpful.

**Comment by zeha** [comment]
Seems like this landed in 2.8? Is this issue now done? /cc @Darlelet  @capflam

**Comment by wtarreau** [comment]
Good point, you're right. I'm closing it now. Thanks for the reminder.
